src/main/java/teammates/common/util/ActivityLogGenerator.java

```diff
@@ -0,0 +1,329 @@
+        
+        builder.withLogMessage(message);
+        
+        return builder.build().generateLogMessage();
+    }
+
+    /**
+     * Generates basic builder for activityLogEntry.
+     * 
+     * @param url URL of the request
+     * @param params parameterMap of the request
+     * @param currUser login information generated by {@link teammates.logic.api.GateKeeper}
+     * @return Builder builder with basic information
+     */
+    private Builder generateBasicActivityLogEntry(String url, Map<String, String[]> params, UserType currUser) {
+        String actionName = getActionNameFromUrl(url);
+        long currTime = System.currentTimeMillis();
+        Builder builder = new Builder(actionName, url, currTime);
+        
+        if (isAutomatedAction(url)) {
+            builder.withLogId(generateLogIdInAutomatedAction(currTime));
+            builder.withUserRole(Const.ActivityLog.ROLE_AUTO);
+        } else if (currUser == null) {
+            builder.withLogId(generateLogIdWithoutGoogleId(params, currTime));
+            builder.withUserGoogleId(Const.ActivityLog.AUTH_NOT_LOGIN);
+        } else {
+            builder.withLogId(generateLogIdWithGoogleId(currUser.id, currTime));
+            builder.withUserGoogleId(currUser.id);
+            updateRoleForLoggedInUser(builder, currUser);
+        }
+        
+        return builder;
+    }
+    
+    private void updateRoleForLoggedInUser(Builder builder, UserType currUser) {
+        if (currUser.isAdmin) {
+            builder.withUserRole(Const.ActivityLog.ROLE_ADMIN);
+            degradeRoleToStudentIfNecessary(builder);
+            degradeRoleToInstructorIfNecessary(builder);
+        } else if (currUser.isInstructor && currUser.isStudent) {
+            builder.withUserRole(Const.ActivityLog.ROLE_INSTRUCTOR);
+            degradeRoleToStudentIfNecessary(builder);
+        } else if (currUser.isStudent) {
+            builder.withUserRole(Const.ActivityLog.ROLE_STUDENT);
+        } else if (currUser.isInstructor) {
+            builder.withUserRole(Const.ActivityLog.ROLE_INSTRUCTOR);
+        } else {
+            builder.withUserRole(Const.ActivityLog.ROLE_UNREGISTERED);
+        }
+    }
+
+    private void degradeRoleToStudentIfNecessary(Builder builder) {
+        if (isStudentPage(builder.getActionName())) {
+            builder.withUserRole(Const.ActivityLog.ROLE_STUDENT);
+        }
+    }
+
+    private void degradeRoleToInstructorIfNecessary(Builder builder) {
+        if (isInstructorPage(builder.getActionName())) {
+            builder.withUserRole(Const.ActivityLog.ROLE_INSTRUCTOR);
+        }
+    }
+
+    private boolean isInstructorPage(String actionName) {
+        return actionName.toLowerCase().startsWith(Const.ActivityLog.PREFIX_INSTRUCTOR_PAGE)
+                || Const.ActionURIs.INSTRUCTOR_FEEDBACK_STATS_PAGE.contains(actionName);
+        // TODO rename the special INSTRUCTOR_FEEDBACK_STATS_PAGE to start with instructor
```

Comment:
This does not directly concern logs. If you want to add this TODO, don't add it here.